### ISSUE TITLE
ERROR: conflicting ColumnTypes while executing meta query on CockroachDB

### ISSUE THREAD

**Issue by kunal-bhatia** [jOOQ/jOOQ#8478]
Exception occurred while running the code generation tool on cockroachdb (postgres)
```
WARNING: SQL exception            : Exception while executing meta query: ERROR: conflicting ColumnTypes: semantic_type:STRING width:0 precision:0 visible_type:NONE  and semantic_type:NAME width:0 precision:0 visible_type:NONE
```
```sql
select
  "tables"."table_schema",
  "tables"."table_name",
  "tables"."specific_name",
  "tables"."table_valued_function",
  "tables"."materialized_view",
  "tables"."description"
from (
  (
    select
      "information_schema"."tables"."table_schema",
      "information_schema"."tables"."table_name",
      "information_schema"."tables"."table_name" as "specific_name",
      false as "table_valued_function",
      false as "materialized_view",
      "pg_catalog"."pg_description"."description"
    from "information_schema"."tables"
      join "pg_catalog"."pg_namespace"
      on "information_schema"."tables"."table_schema" = "pg_catalog"."pg_namespace"."nspname"
      join "pg_catalog"."pg_class"
      on (
        "pg_catalog"."pg_class"."relname" = "information_schema"."tables"."table_name"
        and "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid
      )
      left outer join "pg_catalog"."pg_description"
      on (
        "pg_catalog"."pg_description"."objoid" = "pg_catalog"."pg_class".oid
        and "pg_catalog"."pg_description"."objsubid" = 0
      )
    where (
      "information_schema"."tables"."table_schema" in ('public')
      and ("information_schema"."tables"."table_schema", "information_schema"."tables"."table_name") not in (
        select
          "pg_catalog"."pg_namespace"."nspname",
          "pg_catalog"."pg_class"."relname"
        from "pg_catalog"."pg_class"
          join "pg_catalog"."pg_namespace"
          on "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid
        where "pg_catalog"."pg_class"."relkind" = 'm'
      )
    )
  )
  union all (
    select
      "pg_catalog"."pg_namespace"."nspname",
      "pg_catalog"."pg_class"."relname",
      "pg_catalog"."pg_class"."relname",
      false as "table_valued_function",
      true as "materialized_view",
      "pg_catalog"."pg_description"."description"
    from "pg_catalog"."pg_class"
      join "pg_catalog"."pg_namespace"
      on "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid
      left outer join "pg_catalog"."pg_description"
      on (
        "pg_catalog"."pg_description"."objoid" = "pg_catalog"."pg_class".oid
        and "pg_catalog"."pg_description"."objsubid" = 0
      )
    where (
      "pg_catalog"."pg_namespace"."nspname" in ('public')
      and "pg_catalog"."pg_class"."relkind" = 'm'
    )
  )
  union all (
    select
      "information_schema"."routines"."routine_schema",
      "information_schema"."routines"."routine_name",
      "information_schema"."routines"."specific_name",
      true as "table_valued_function",
      false as "materialized_view",
      ''
    from "information_schema"."routines"
      join "pg_catalog"."pg_namespace"
      on "information_schema"."routines"."specific_schema" = "pg_catalog"."pg_namespace"."nspname"
      join "pg_catalog"."pg_proc"
      on (
        "pg_catalog"."pg_proc"."pronamespace" = "pg_catalog"."pg_namespace".oid
        and (("pg_catalog"."pg_proc"."proname" || '_') || cast("pg_catalog"."pg_proc".oid as varchar)) = "information_schema"."routines"."specific_name"
      )
    where (
      "information_schema"."routines"."routine_schema" in ('public')
      and "pg_catalog"."pg_proc"."proretset"
    )
  )
) as "tables"
order by
  1,
  2
```

```
Apr 01, 2019 1:20:06 PM org.jooq.tools.JooqLogger error
SEVERE: Error while fetching tables
org.jooq.exception.DataAccessException: SQL [select "tables"."table_schema", "tables"."table_name", "tables"."specific_name", "tables"."table_valued_function", "tables"."materialized_view", "tables"."description" from ((select "information_schema"."tables"."table_schema", "information_schema"."tables"."table_name", "information_schema"."tables"."table_name" as "specific_name", false as "table_valued_function", false as "materialized_view", "pg_catalog"."pg_description"."description" from "information_schema"."tables" join "pg_catalog"."pg_namespace" on "information_schema"."tables"."table_schema" = "pg_catalog"."pg_namespace"."nspname" join "pg_catalog"."pg_class" on ("pg_catalog"."pg_class"."relname" = "information_schema"."tables"."table_name" and "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid) left outer join "pg_catalog"."pg_description" on ("pg_catalog"."pg_description"."objoid" = "pg_catalog"."pg_class".oid and "pg_catalog"."pg_description"."objsubid" = ?) where ("information_schema"."tables"."table_schema" in (?) and ("information_schema"."tables"."table_schema", "information_schema"."tables"."table_name") not in (select "pg_catalog"."pg_namespace"."nspname", "pg_catalog"."pg_class"."relname" from "pg_catalog"."pg_class" join "pg_catalog"."pg_namespace" on "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid where "pg_catalog"."pg_class"."relkind" = 'm'))) union all (select "pg_catalog"."pg_namespace"."nspname", "pg_catalog"."pg_class"."relname", "pg_catalog"."pg_class"."relname", false as "table_valued_function", true as "materialized_view", "pg_catalog"."pg_description"."description" from "pg_catalog"."pg_class" join "pg_catalog"."pg_namespace" on "pg_catalog"."pg_class"."relnamespace" = "pg_catalog"."pg_namespace".oid left outer join "pg_catalog"."pg_description" on ("pg_catalog"."pg_description"."objoid" = "pg_catalog"."pg_class".oid and "pg_catalog"."pg_description"."objsubid" = ?) where ("pg_catalog"."pg_namespace"."nspname" in (?) and "pg_catalog"."pg_class"."relkind" = 'm')) union all (select "information_schema"."routines"."routine_schema", "information_schema"."routines"."routine_name", "information_schema"."routines"."specific_name", true as "table_valued_function", false as "materialized_view", '' from "information_schema"."routines" join "pg_catalog"."pg_namespace" on "information_schema"."routines"."specific_schema" = "pg_catalog"."pg_namespace"."nspname" join "pg_catalog"."pg_proc" on ("pg_catalog"."pg_proc"."pronamespace" = "pg_catalog"."pg_namespace".oid and (("pg_catalog"."pg_proc"."proname" || ?) || cast("pg_catalog"."pg_proc".oid as varchar)) = "information_schema"."routines"."specific_name") where ("information_schema"."routines"."routine_schema" in (?) and "pg_catalog"."pg_proc"."proretset"))) as "tables" order by 1, 2]; ERROR: conflicting ColumnTypes: semantic_type:STRING width:0 precision:0 visible_type:NONE  and semantic_type:NAME width:0 precision:0 visible_type:NONE
	at org.jooq_3.11.2.POSTGRES.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2380)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:811)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:364)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:323)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2700)
	at org.jooq.meta.postgres.PostgresDatabase.getTables0(PostgresDatabase.java:433)
	at org.jooq.meta.AbstractDatabase.getTables(AbstractDatabase.java:1274)
	at org.jooq.codegen.JavaGenerator.generateSchemaIfEmpty(JavaGenerator.java:419)
	at org.jooq.codegen.JavaGenerator.generateCatalogIfEmpty(JavaGenerator.java:405)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:379)
	at org.jooq.codegen.GenerationTool.run(GenerationTool.java:750)
	at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:220)
	at org.jooq.codegen.GenerationTool.main(GenerationTool.java:192)
Caused by: org.postgresql.util.PSQLException: ERROR: conflicting ColumnTypes: semantic_type:STRING width:0 precision:0 visible_type:NONE  and semantic_type:NAME width:0 precision:0 visible_type:NONE
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2433)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2178)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:306)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:155)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:144)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:3479)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:268)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:350)
	... 10 more
```

**Versions:**

- jOOQ - 3.11.2
- Java - 11.0.1
- Database (include vendor) - CockroachDB 2.1.4
- OS - Mac OS X
- JDBC Driver (PostgreSQL Version) -  42.2.2

**Comment by lukaseder** [comment]
Thank you very much for your report. Do note we currently do not support CockroachDB and we have no way of integration testing any potential fix.

Having said so, we can still try to work around this problem if you can help finding the cause and providing an alternative query syntax that still does the same thing as the original query. This means, we'll have to rely on your help on patching this for you. How could this query be rewritten to work on CockroachDB?

**Comment by knz** [comment]
The problem you are encountering is that CockroachDB uses a stronger type system than PostgreSQL. As the error indicates, it is not valid in CockroachDB to compare a value of type NAME directly with a value of type STRING in a UNION operator without a cast.

The error originates because the "name" columns in `information_schema.tables` have type STRING/TEXT, whereas the column `relname` in `pg_catalog.pg_class` has type NAME.

The solution is simple it is simply to add the appropriate cast. Change these 3 lines:

```sql
                            SELECT pg_catalog.pg_namespace.nspname,
                                   pg_catalog.pg_class.relname,
                                   pg_catalog.pg_class.relname,
```

To this:

```sql
                            SELECT pg_catalog.pg_namespace.nspname::STRING,
                                   pg_catalog.pg_class.relname::STRING,
                                   pg_catalog.pg_class.relname::STRING,
```

The final result is:

```sql
SELECT tables.table_schema,
         tables.table_name,
         tables.specific_name,
         tables.table_valued_function,
         tables.materialized_view,
         tables.description
    FROM (
               (
                SELECT information_schema.tables.table_schema,
                       information_schema.tables.table_name,
                       information_schema.tables.table_name AS specific_name,
                       false AS table_valued_function,
                       false AS materialized_view,
                       pg_catalog.pg_description.description
                  FROM information_schema.tables
                       JOIN pg_catalog.pg_namespace
                       ON information_schema.tables.table_schema = pg_catalog.pg_namespace.nspname
                       JOIN pg_catalog.pg_class
                       ON pg_catalog.pg_class.relname = information_schema.tables.table_name
                      AND pg_catalog.pg_class.relnamespace = pg_catalog.pg_namespace.oid
                       LEFT JOIN pg_catalog.pg_description
                       ON pg_catalog.pg_description.objoid = pg_catalog.pg_class.oid
                      AND pg_catalog.pg_description.objsubid = 0
                 WHERE information_schema.tables.table_schema IN ('public',)
                   AND (information_schema.tables.table_schema, information_schema.tables.table_name)
                       NOT IN (
                            SELECT pg_catalog.pg_namespace.nspname, pg_catalog.pg_class.relname
                              FROM pg_catalog.pg_class
                                   JOIN pg_catalog.pg_namespace
                                   ON pg_catalog.pg_class.relnamespace = pg_catalog.pg_namespace.oid
                             WHERE pg_catalog.pg_class.relkind = 'm'
                        )
               )
              UNION ALL  (
                            SELECT pg_catalog.pg_namespace.nspname::STRING,
                                   pg_catalog.pg_class.relname::STRING,
                                   pg_catalog.pg_class.relname::STRING,
                                   false AS table_valued_function,
                                   true AS materialized_view,
                                   pg_catalog.pg_description.description
                              FROM pg_catalog.pg_class
                                   JOIN pg_catalog.pg_namespace
                                   ON pg_catalog.pg_class.relnamespace = pg_catalog.pg_namespace.oid
                                   LEFT JOIN pg_catalog.pg_description
                                   ON pg_catalog.pg_description.objoid = pg_catalog.pg_class.oid
                                  AND pg_catalog.pg_description.objsubid = 0
                             WHERE pg_catalog.pg_namespace.nspname IN ('public',)
                               AND pg_catalog.pg_class.relkind = 'm'
                         )
             UNION ALL  (
                            SELECT information_schema.routines.routine_schema,
                                   information_schema.routines.routine_name,
                                   information_schema.routines.specific_name,
                                   true AS table_valued_function,
                                   false AS materialized_view,
                                   ''
                              FROM information_schema.routines
                                   JOIN pg_catalog.pg_namespace
                                   ON information_schema.routines.specific_schema
                                      = pg_catalog.pg_namespace.nspname
                                   JOIN pg_catalog.pg_proc
                                   ON pg_catalog.pg_proc.pronamespace = pg_catalog.pg_namespace.oid
                                  AND (
                                        pg_catalog.pg_proc.proname
                                        || '_'
                                        || CAST(pg_catalog.pg_proc.oid AS VARCHAR)
                                      )
                                      = information_schema.routines.specific_name
                             WHERE information_schema.routines.routine_schema IN ('public',)
                               AND pg_catalog.pg_proc.proretset
                        )
         ) AS tables
ORDER BY 1, 2;
```

**Comment by knz** [comment]
You can also use a `::TEXT` cast instead of `::STRING` so that the resulting query remains compatible with both CockroachDB and PostgreSQL.

**Comment by lukaseder** [comment]
Thanks a lot for your hint. `::TEXT` is definitely preferrable, or even `::VARCHAR`, as the `information_schema` types are of type `sql_identifier`:

```sql
CREATE DOMAIN information_schema.sql_identifier AS varchar
	COLLATE default;
```

Does the query still work when using casts to `VARCHAR`, in CockroachDB? I wouldn't want to risk a regression where `TEXT` and `VARCHAR` are incompatible types...

**Comment by lukaseder** [comment]
Hmm, tried back to PostgreSQL 9.3. This still works:

```sql
select 'a'::varchar
union all
select 'a'::text
```

So, I guess the ``::TEXT` cast is safe

**Comment by knz** [comment]
`varchar` and `text` are compatible **in PostgreSQL** because pg really accepts to compare anything with anything that's text-like.

The sensitivity is in CockroachDB. 

In this specific case, I recommend a cast to `::varchar` on both sides (or `::text`, it does not matter, what matters is that they're the same): ensure everything goes to `varchar` and avoid relying on equivalence between types without an explicit cast.

**Comment by lukaseder** [comment]
> varchar and text are compatible in PostgreSQL because pg really accepts to compare anything with anything that's text-like.

There's always a caveat ;-) Try this in plain old PostgreSQL:

```sql
select array['a']::varchar[] = array['a']::text[]
```

So, I'll go with `::varchar`. The only change I've noticed is a change in sort order, which isn't too relevant.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12. Backport scheduled for 3.11.11 (#8485).

**Comment by lukaseder** [comment]
Thanks again for reporting. We'll backport this fix to 3.11.11 soon. If you could integration test this on your side, this would be great, as we currently don't have a means to test this query on CockroachDB. Here's the patch for this change (applied to 3.12, should be similar in 3.11):

https://github.com/jOOQ/jOOQ/commit/c0ada90f6e11c6bac2321b465162068dde21a42f.patch

Let me know if I can help with applying the patch.

**Comment by kunal-bhatia** [comment]
Thanks a lot for the same, I will do some integration testing with it and let you know in case of issues.
